packages/google-auth/src/authFlow.ts

```diff
@@ -0,0 +1,153 @@
+import { html } from 'hono/html'
+import { HTTPException } from 'hono/http-exception'
+
+import type { GoogleErrorResponse, GoogleTokenResponse, GoogleUser, GoogleTokenInfoResponse, Token } from './types'
+import { toQueryParams } from './utils/toQueryParams'
+
+type GoogleAuthFlow = {
+  client_id: string
+  client_secret: string
+  redirect_uri: string
+  response_type: 'code' | 'token'
+  code: string | undefined
+  token: Token | undefined
+  scope: string[]
+  include_granted_scopes?: boolean
+  state?: string
+  login_hint?: string
+  prompt?: 'none' | 'consent' | 'select_account'
+}
+
+export class AuthFlow {
+  client_id: string
+  client_secret: string
+  redirect_uri: string
+  response_type: 'code' | 'token'
+  code: string | undefined
+  token: Token | undefined
+  scope: string[]
+  include_granted_scopes: boolean | undefined
+  state: string | undefined
+  login_hint: string | undefined
+  prompt: 'none' | 'consent' | 'select_account' | undefined
+  user: Partial<GoogleUser> | undefined
+  granted_scopes: string[] | undefined
+
+  constructor({
+    client_id, client_secret, redirect_uri,
+    response_type,include_granted_scopes,
+    login_hint, prompt, scope, state, code, token
+  }: GoogleAuthFlow) {
+    this.client_id = client_id
+    this.client_secret = client_secret
+    this.redirect_uri = redirect_uri
+    this.response_type = response_type
+    this.include_granted_scopes = include_granted_scopes
+    this.login_hint = login_hint
+    this.prompt = prompt
+    this.scope = scope
+    this.state = state
+    this.code = code
+    this.token = token
+    this.user = undefined
+    this.granted_scopes = undefined
+
+    if (
+      this.client_id === undefined || this.client_secret === undefined ||
+      this.scope === undefined || this.response_type === undefined
+    ) {
+      throw new HTTPException(401, { message: 'Required parameters were not found. Please provide them to proceed.' })
+    }
+  }
+
+  redirect() {
+    const parsedOptions = toQueryParams({
+      response_type: this.response_type,
+      redirect_uri: this.redirect_uri,
+      client_id: this.client_id,
+      include_granted_scopes: this.include_granted_scopes,
+      scope: this.scope.join(' '),
+      state: this.state
+    })
+    const url = `https://accounts.google.com/o/oauth2/v2/auth?${parsedOptions}`
+
+    if (this.response_type === 'token') {
+      return html`
```

Comment:
Any reasons to redirect users client side, and not just use `Location` header?

Reply:
Could't get the params because of the hash on the redirect uri from google. I'll use only the code flow now so this will be removed.